### PR TITLE
feat(python,rust): frame-level `cast` support

### LINKED ISSUE #9959: Add `.cast` as DataFrame method
### Problem description

It would be useful to add .cast as a method on the dataframe class which takes a dictionary as input with full or partial schema and their polars dtypes. 

Example:

```python
schema = {"foo": pl.Uint64, "bar": pl.List}

df = df.cast(schema)

```

### PR DESCRIPTION
**Pull request by alexander-beedie**
Closes  #9959.

* Compact syntax can help reduce common boilerplate.
* Supports both `DataFrame` and `LazyFrame`.
* Can target the columns to cast using...
  * selectors [^1]
  * named columns  
  * a single target dtype
  * partial/full schemas from other frames
  
## Example

```python
from datetime import date
import polars as pl

df = pl.DataFrame({
    "foo": [1, 2, 3],
    "bar": [6.0, 7.0, 8.0],
    "baz": [date(2020,1,2), date(2021,3,4), date(2022,5,6)],
})
```
Specific cols:
```python
df.cast( {"foo":pl.Float32, "bar":pl.UInt8, "baz":pl.Datetime("ms")} )
# ┌─────┬─────┬─────────────────────┐
# │ foo ┆ bar ┆ baz                 │
# │ --- ┆ --- ┆ ---                 │
# │ f32 ┆ u8  ┆ datetime[ms]        │
# ╞═════╪═════╪═════════════════════╡
# │ 1.0 ┆ 6   ┆ 2020-01-02 00:00:00 │
# │ 2.0 ┆ 7   ┆ 2021-03-04 00:00:00 │
# │ 3.0 ┆ 8   ┆ 2022-05-06 00:00:00 │
# └─────┴─────┴─────────────────────┘
```
Selectors:
```python
df.cast( {cs.numeric(): pl.Float32} )
# ┌─────┬─────┬────────────┐
# │ foo ┆ bar ┆ baz        │
# │ --- ┆ --- ┆ ---        │
# │ f32 ┆ f32 ┆ date       │
# ╞═════╪═════╪════════════╡
# │ 1.0 ┆ 6.0 ┆ 2020-01-02 │
# │ 2.0 ┆ 7.0 ┆ 2021-03-04 │
# │ 3.0 ┆ 8.0 ┆ 2022-05-06 │
# └─────┴─────┴────────────┘
```
Single target dtype:
```python
df.cast( pl.Utf8 ).to_dict( False )
# {'foo': ['1', '2', '3'],
#  'bar': ['6.0', '7.0', '8.0'],
#  'baz': ['2020-01-02', '2021-03-04', '2022-05-06']}
```

[^1]: Polars selectors: https://pola-rs.github.io/polars/py-polars/html/reference/selectors.html

### CODE REVIEW

File: py-polars/tests/unit/test_selectors.py
Comment (reswqa) [inline py-polars/tests/unit/test_selectors.py:342]: I guess we need wrap it with `repr`.

File: py-polars/tests/unit/test_selectors.py
Comment (reswqa) [inline py-polars/tests/unit/test_selectors.py:343]: Not sure if  `include_categorical` should take into `cs.string`'s repr. If yes, perhaps we should pass it to `_selector_proxy_`'s `__init__`?

**Comment by reswqa** [comment]
Thanks for introducing this, really helpful feature! This change overall look good to me, I just left nit comments related to a unit test.

File: py-polars/tests/unit/test_selectors.py
Comment (alexander-beedie) [reply py-polars/tests/unit/test_selectors.py:342]: Ouch... indeed :)

File: py-polars/tests/unit/test_selectors.py
Comment (alexander-beedie) [reply py-polars/tests/unit/test_selectors.py]: We already do - that's what `parameters` is for, in order to show what the higher-level selector was instantiated with.

**Comment by ritchie46** [comment]
I think this is nice. I am only a bit hesitant about this being implemented in python. I'd rather have this available on the rust `LazyFrame` and dispatch to that.

**Comment by alexander-beedie** [comment]
> I think this is nice. I am only a bit hesitant about this being implemented in python. I'd rather have this available on the rust `LazyFrame` and dispatch to that.

Sure; let me put my Rust hat on and see how it looks... ;)

**Comment by alexander-beedie** [comment]
> I'd rather have this available on the rust `LazyFrame` and dispatch to that.

Done; there's now a `cast` on the Rust-side LazyFrame with the following signature:
```rust
fn cast(&self, dtypes: HashMap<&str, Wrap<DataType>>)
```
The selector and dtype options on the Python side get expanded into an appropriate map before passing down, since overloading isn't really Rust's strong suit, but this is definitely looking better for feature equivalence, etc 👍

**Comment by stinodego** [comment]
@alexander-beedie I don't see any changes on the Rust side - did you forget to push your latest changes?

**Comment by alexander-beedie** [comment]
> @alexander-beedie I don't see any changes on the Rust side - did you forget to push your latest changes?

No, they're there - have a look: https://github.com/pola-rs/polars/pull/10504/commits/2e3dc06108643561d355e611ec02fadaebbae049  ;)

File: py-polars/polars/lazyframe/frame.py
Comment (stinodego) [inline py-polars/polars/lazyframe/frame.py:2065]: I can imagine we want this functionality available on the Rust side as well, though obviously it would need to be a different method. `cast_single_dtype`?

Not sure how far the parity between Rust and Python needs to go (I'd classify this one as a gray area).

File: py-polars/polars/lazyframe/frame.py
Comment (stinodego) [inline py-polars/polars/lazyframe/frame.py:2076]: My brain always has difficulty parsing this ternary syntax... isn't this more readable?

```suggestion
                if isinstance(c, str):
                    cast_map[c] = dtype
                else:
                    cast_map.update({x: dtype for x in expand_selector(self, c)})
```

File: py-polars/polars/testing/asserts.py
Comment (stinodego) [inline py-polars/polars/testing/asserts.py:575]: I don't think we want this in the public API, at least for now. You can define it in the test module you're using it in or maybe in `conftest.py`. I did something similar for `assert_expr_equal` somewhere.

**Review by stinodego — CHANGES_REQUESTED** [review]
Just looking at the doc examples makes me excited to have this functionality available!

I left a few remarks, mostly details.

I'd like to defer to @ritchie46 for reviewing the details of the Rust side (though it looks fine to me).

File: py-polars/polars/lazyframe/frame.py
Comment (alexander-beedie) [reply py-polars/polars/lazyframe/frame.py:2067]: No reason at all - in fact I have no clue why it's there; fixing...

File: py-polars/polars/lazyframe/frame.py
Comment (alexander-beedie) [reply py-polars/polars/lazyframe/frame.py:2076]: Not to me; I actually find the ternary construct cleaner there, but I guess that's down to personal taste and/or which languages you're most attuned to - I can change it though 😅

Comment (stinodego) [reply py-polars/polars/lazyframe/frame.py]: In that case, let's leave it as-is!

File: py-polars/polars/lazyframe/frame.py
Comment (alexander-beedie) [reply py-polars/polars/lazyframe/frame.py:2065]: > obviously it would need to be a different method. cast_single_dtype?

Or `cast_all`? But yes, I wasn't convinced on the Rust side if we want to eat the extra API surface given the relative ease of `select( all().cast(Utf8) )`.

File: py-polars/polars/testing/asserts.py
Comment (alexander-beedie) [reply py-polars/polars/testing/asserts.py:575]: Sure; I'll move somewhere else (it's used in more than one test file, but it doesn't need to be part of our own public API, quite right)

File: py-polars/polars/lazyframe/frame.py
Comment (stinodego) [reply py-polars/polars/lazyframe/frame.py:2065]: I like `cast_all`!

Basically, if the functionality is worth having on the Python side, it's worth having on the Rust side. Mostly, anyway.

In this case, I think it's worth having.

Comment (alexander-beedie) [reply py-polars/polars/lazyframe/frame.py:2065]: "Sold" - I'll throw it in :)

Comment (alexander-beedie) [reply py-polars/polars/lazyframe/frame.py]: Done! ✌️  (Also just rebased onto the latest code to pick up the `wasm` fix for CI)

**Comment by alexander-beedie** [comment]
Ok, should have addressed all the comments! Implementations for `cast` and `cast_all` can be found on `LazyFrame` and `PyLazyFrame` in Rust, and we now dispatch to those (after having expanded any Python-side selectors).

**Review by stinodego — APPROVED** [review]
This looks good to me! Though I can imagine more seasoned Rustaceans would write the `cast` method using some generic mapping trait instead of a `PlHashMap`. Maybe @orlp or @ritchie46 want to jump in here?

In any case, has my blessing!

**Comment by orlp** [comment]
> This looks good to me! Though I can imagine more seasoned Rustaceans would write the cast method using some generic mapping trait instead of a PlHashMap. Maybe @orlp or @ritchie46 want to jump in here?

I think this is perfectly fine.

**Pull request merged**
